File: android/src/main/java/com/iconshot/detonator/request/Request.java
package com.iconshot.detonator.request;

import com.iconshot.detonator.Detonator;
import com.iconshot.detonator.renderer.Edge;

public abstract class Request<T> {
    protected Detonator detonator;

    protected int id;
    protected Integer componentId;
    protected String data;

    public Request(Detonator detonator, IncomingRequest incomingRequest) {
        this.detonator = detonator;

        this.id = incomingRequest.id;
        this.componentId = incomingRequest.componentId;
        this.data = incomingRequest.data;
    }

    protected T decodeData(Class<T> DataClass) {
        if (data == null) {
            return null;
        }

        return detonator.decode(data, DataClass);
    }

    protected Edge getComponentEdge() {
        if (componentId == null) {
            return null;
        }

        return detonator.getEdge(componentId);
    }

    public abstract void run();

    protected void end() {
        end(null);
    }

    protected void end(Object data) {
        OutgoingResponse response = new OutgoingResponse();

        response.id = id;
        response.data = data;

        emit(response);
    }

    protected void error(Exception exception) {
        OutgoingResponse response = new OutgoingResponse();

        Error error = new Error();

        error.message = exception.getMessage();

        response.id = id;
        response.error = error;

        emit(response);
    }

    private void emit(OutgoingResponse response) {
        detonator.emit("response", response);
    }

    public static class IncomingRequest {
        public int id;
        public String name;
        public Integer componentId;
        public String data;
    }

    public static class OutgoingResponse {
        public int id;
        public Object data;
        public Error error;
    }

    public static class Error {
        public String message;
    }
}

File: ios/Detonator/Sources/Detonator/Request/Request.swift
import UIKit

open class Request {
    public let detonator: Detonator
    
    let id: Int
    let componentId: Int?
    let data: String?
    
    required public init(_ detonator: Detonator, incomingRequest: IncomingRequest) {
        self.detonator = detonator
        
        self.id = incomingRequest.id
        self.componentId = incomingRequest.componentId
        self.data = incomingRequest.data
    }
    
    public func decodeData<T: Decodable>() -> T? {
        guard let data = data else {
            return nil
        }
        
        return detonator.decode(data)
    }
    
    public func getComponentEdge() -> Edge? {
        if componentId == nil {
            return nil
        }
        
        return detonator.getEdge(edgeId: componentId!)
    }
    
    open func run() {}
    
    public func end(data: Encodable? = nil) {
        let response = OutgoingResponse(id: id, data: data, error: nil)
        
        emitResponse(response: response)
    }
    
    public func error(message: String) {
        let tmpError = OutgoingResponse.Error(message: message)
        
        let response = OutgoingResponse(id: id, data: nil, error: tmpError)
        
        emitResponse(response: response)
    }
    
    public func error(error: Error) {
        self.error(message: error.localizedDescription)
    }
    
    private func emitResponse(response: OutgoingResponse) {
        detonator.emit("com.iconshot.detonator.request.response", response)
    }
    
    public struct IncomingRequest: Decodable {
        let id: Int
        let name: String
        let componentId: Int?
        let data: String?
    }
    
    public struct OutgoingResponse: Encodable {
        let id: Int
        let data: NullEncodable<AnyEncodable>?
        let error: NullEncodable<Error>?
        
        init(id: Int, data: Encodable?, error: Error?) {
            self.id = id
            
            if let data = data {
                self.data = NullEncodable(AnyEncodable(data))
            } else {
                self.data = NullEncodable(nil)
            }
            
            self.error = NullEncodable(error)
        }
        
        struct Error: Encodable {
            let message: String
        }
    }
}
